azurerm/resource_arm_azuread_application.go

```diff
@@ -0,0 +1,214 @@
+	return &schema.Resource{
+		Create: resourceArmActiveDirectoryApplicationCreate,
+		Read:   resourceArmActiveDirectoryApplicationRead,
+		Update: resourceArmActiveDirectoryApplicationUpdate,
+		Delete: resourceArmActiveDirectoryApplicationDelete,
+		Importer: &schema.ResourceImporter{
+			State: schema.ImportStatePassthrough,
+		},
+
+		Schema: map[string]*schema.Schema{
+			"name": {
+				Type:     schema.TypeString,
+				Required: true,
+			},
+
+			"homepage": {
+				Type:     schema.TypeString,
+				Optional: true,
+				Computed: true,
+			},
+
+			"identifier_uris": {
+				Type:     schema.TypeList,
+				Optional: true,
+				Computed: true,
+				MinItems: 1,
+				Elem: &schema.Schema{
+					Type: schema.TypeString,
+				},
+			},
+
+			"reply_urls": {
+				Type:     schema.TypeList,
+				Optional: true,
+				Computed: true,
+				MinItems: 1,
+				Elem: &schema.Schema{
+					Type: schema.TypeString,
+				},
+			},
+
+			"available_to_other_tenants": {
+				Type:     schema.TypeBool,
+				Optional: true,
+				Computed: true,
+			},
+
+			"oauth2_allow_implicit_flow": {
+				Type:     schema.TypeBool,
+				Optional: true,
+				Computed: true,
+			},
+
+			"application_id": {
+				Type:     schema.TypeString,
+				Computed: true,
+			},
+		},
+	}
+}
+
+func resourceArmActiveDirectoryApplicationCreate(d *schema.ResourceData, meta interface{}) error {
+	client := meta.(*ArmClient).applicationsClient
+	ctx := meta.(*ArmClient).StopContext
+
+	name := d.Get("name").(string)
+	multitenant := d.Get("available_to_other_tenants").(bool)
+
+	properties := graphrbac.ApplicationCreateParameters{
+		DisplayName:             &name,
+		Homepage:                expandAzureRmActiveDirectoryApplicationHomepage(d, name),
+		IdentifierUris:          expandAzureRmActiveDirectoryApplicationIdentifierUris(d, name),
+		ReplyUrls:               expandAzureRmActiveDirectoryApplicationReplyUrls(d, name),
+		AvailableToOtherTenants: utils.Bool(multitenant),
+	}
+
+	if v, ok := d.GetOk("oauth2_allow_implicit_flow"); ok {
+		properties.Oauth2AllowImplicitFlow = utils.Bool(v.(bool))
+	}
+
+	app, err := client.Create(ctx, properties)
+	if err != nil {
+		return err
+	}
+
+	d.SetId(*app.ObjectID)
+
+	return resourceArmActiveDirectoryApplicationRead(d, meta)
+}
+
+func resourceArmActiveDirectoryApplicationRead(d *schema.ResourceData, meta interface{}) error {
+	client := meta.(*ArmClient).applicationsClient
+	ctx := meta.(*ArmClient).StopContext
+
+	resp, err := client.Get(ctx, d.Id())
+	if err != nil {
+		if utils.ResponseWasNotFound(resp.Response) {
+			log.Printf("[DEBUG] Azure AD Application ID %q was not found - removing from state", d.Id())
+			d.SetId("")
+			return nil
+		}
+
+		return fmt.Errorf("Error loading Azure AD Application %q: %+v", d.Id(), err)
+	}
+
+	d.Set("name", resp.DisplayName)
+	d.Set("application_id", resp.AppID)
+	d.Set("homepage", resp.Homepage)
+	d.Set("identifier_uris", resp.IdentifierUris)
+	d.Set("reply_urls", resp.ReplyUrls)
```

Comment:
given these are complex objects - I'm going to push a commit to add a flatten function for each of these and check for errors when setting them